src/main/java/com/google/finapp/SpannerDaoJDBCImpl.java

```diff
@@ -0,0 +1,197 @@
+    this.connectionUrl = String.format(
+        "jdbc:cloudspanner://localhost:9010/projects/%s/instances/%s/databases/%s;usePlainText=true",
+        projectId, instanceId, databaseId);
+    // use this URL to connect to Cloud Spanner
+    // this.connectionUrl = String.format(
+    //     "jdbc:cloudspanner:/projects/%s/instances/%s/databases/%s",
+    //     projectId, instanceId, databaseId);
+  }
+
+  public void createCustomer(ByteArray customerId, String name, String address) throws SQLException {
+    try (Connection connection = DriverManager.getConnection(this.connectionUrl)) {
+      try (PreparedStatement ps =
+          connection.prepareStatement(
+              "INSERT INTO Customer\n"
+                  + "(CustomerId, Name, Address)\n"
+                  + "VALUES\n"
+                  + "(?, ?, ?)")) {
+        ps.setBytes(1, customerId.toByteArray());
+        ps.setString(2, name);
+        ps.setString(3, address);
+        ps.executeUpdate();
+        System.out.printf("Customer created: %s\n", name);
+      }
+    }
+  }
+
+
+  public void createAccount(
+      ByteArray accountId, AccountType accountType, AccountStatus accountStatus, BigDecimal balance)
+      throws SQLException {
+    try (Connection connection = DriverManager.getConnection(this.connectionUrl)) {
+      try (PreparedStatement ps =
+          connection.prepareStatement(
+              "INSERT INTO Account\n"
+                  + "(AccountId, AccountType, AccountStatus, Balance, CreationTimestamp)\n"
+                  + "VALUES\n"
+                  + "(?, ?, ?, ?, PENDING_COMMIT_TIMESTAMP())")) {
+        ps.setBytes(1, accountId.toByteArray());
+        ps.setInt(2, accountType.getNumber());
+        ps.setInt(3, accountStatus.getNumber());
+        ps.setBigDecimal(4, balance);
+        ps.executeUpdate();
+        System.out.printf("Account created with balance %s\n", balance.toString());
+      }
+    }
+  }
+
+  public void addAccountForCustomer(
+      ByteArray customerId, ByteArray accountId, ByteArray roleId, String roleName)
+      throws SQLException {
+    try (Connection connection = DriverManager.getConnection(this.connectionUrl)) {
+      try (PreparedStatement ps =
+          connection.prepareStatement(
+              "INSERT INTO CustomerRole\n"
+                  + "(CustomerId, AccountId, RoleId, Role)\n"
+                  + "VALUES\n"
+                  + "(?, ?, ?, ?)")) {
+        ps.setBytes(1, customerId.toByteArray());
+        ps.setBytes(2, accountId.toByteArray());
+        ps.setBytes(3, roleId.toByteArray());
+        ps.setString(4, roleName);
+        ps.executeUpdate();
+        System.out.printf("New role created: %s\n", roleName);
+      }
+    }
+  }
+
+
+  public void moveAccountBalance(
+      ByteArray fromAccountId, ByteArray toAccountId, BigDecimal amount)
+      throws SQLException {
+    try (Connection connection = DriverManager.getConnection(this.connectionUrl)) {
+      // begin transaction
```

Comment:
Is this comment necessary ? There's only one transaction which spans the entire method.

Reply:
The reason I added this comment is to explain that `connection.setAutoCommit(false)` is the action that sets up the transaction, since you turn off auto-commit in order to make the following `execute`s part of a transaction. However, I guess the target audience can probably understand that turning off auto-commit and having the `commit` at the end makes it a transaction.